[Core] Tear apart `from_pretrained()` of `DiffusionPipeline`

sayakpaul opened the pull request:
# What does this PR do?

As discussed [internally](https://huggingface.slack.com/archives/C065E480NN9/p1721883418866279). 

## Approach taken 

Broadly, I have tried to collate logically meaningful code blocks together and moved them to mini functions to promote readability better. 

1. For errors and warnings, I have maintained the explicit nature i.e., raise them from `from_pretrained()` as much as possible and as early as possible. 
2. For functions like `_update_init_kwargs_with_connected_pipeline()`, I decided to just pass a copy of the `kwargs` as this way we reduce LoCs quite a bit. I can follow a similar approach for other methods like `load_sub_model()`, `_determine_device_map()`, too but I would like to get opinions first. 

Some comments are in line as well.

sayakpaul commented on src/diffusers/pipelines/pipeline_utils.py:
```diff
@@ -787,26 +767,18 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
         expected_modules, optional_kwargs = cls._get_signature_keys(pipeline_class)
         passed_class_obj = {k: kwargs.pop(k) for k in expected_modules if k in kwargs}
         passed_pipe_kwargs = {k: kwargs.pop(k) for k in optional_kwargs if k in kwargs}
-
         init_dict, unused_kwargs, _ = pipeline_class.extract_init_dict(config_dict, **kwargs)
```
Why not move these to a function, too? 

That is because there as many arguments in conversation here and I think it's better to keep them as is.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_8967). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

sayakpaul commented:
@DN6 would appreciate a first review of this.

sayakpaul commented:
@DN6 a gentle ping.

DN6 commented on src/diffusers/pipelines/pipeline_utils.py:
```diff
@@ -900,68 +867,25 @@ def load_module(name, value):
             )
 
+        # 9. Potentially add passed objects if expected
+        init_kwargs = _ensure_all_expected_modules_presence(
```
Would name it something like:  `_check_and_update_init_kwargs_for_missing_modules`

sayakpaul replied:
Agreed. Updated in af0769408.

DN6 commented on src/diffusers/pipelines/pipeline_loading_utils.py:
```diff
@@ -749,3 +750,157 @@ def _fetch_class_library_tuple(module):
+def _maybe_raise_warning_for_inpainting(pipeline_class, pretrained_model_name_or_path: str, config: dict):
+    if pipeline_class.__name__ == "StableDiffusionInpaintPipeline" and version.parse(
+        version.parse(config["_diffusers_version"]).base_version
+    ) <= version.parse("0.5.1"):
```
Won't this be raised only is diffusers is `<=0.5.1`? Can't we just get rid of it completely?

sayakpaul replied:
That seems reasonable but I would slightly prefer to keep it even though it's highly unlikely that someone is still using a version that old.

DN6 replied:
Sure. We can remove in v1.0.0

DN6 reviewed (commented):
Looking very nice. Just some minor comments.

DN6 reviewed (approved):
cc: @yiyixuxu To also take a look here.

yiyixuxu commented on src/diffusers/pipelines/pipeline_utils.py:
```diff
@@ -724,59 +734,28 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
-            custom_class_name = config_dict["_class_name"][1]
-
-        pipeline_class = _get_pipeline_class(
+        custom_pipeline, pipeline_class = _determine_pipeline_class(
```
can we just add a method to get the custom pipeline class here so we don't have to have a small function nested in another?
```suggestion
        custom_pipeline = _get_custom_pipeline_class(..)
        pipeline_class = _get_pipeline_class(..)
```

sayakpaul replied:
Done in 31436906c.

yiyixuxu commented on src/diffusers/pipelines/pipeline_utils.py:
```diff
@@ -724,59 +734,28 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
-                " the {StableDiffusionInpaintPipelineLegacy} class and will likely remove it in version 1.0.0."
-            )
-            deprecate("StableDiffusionInpaintPipelineLegacy", "1.0.0", deprecation_message, standard_warn=False)
+        _maybe_raise_warning_for_inpainting(
```
let's either add a comment here explaining what we deprecate here

sayakpaul replied:
Done in 50bb47006.

yiyixuxu commented on src/diffusers/pipelines/pipeline_utils.py:
```diff
@@ -799,14 +777,7 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
-            return True
-
-        init_dict = {k: v for k, v in init_dict.items() if load_module(k, v)}
+        init_dict = _filter_null_components(init_dict=init_dict, passed_class_objs=passed_class_obj)
```
I think it is better without the method

sayakpaul replied:
Done in 1ddba3e6f.

yiyixuxu commented on src/diffusers/pipelines/pipeline_utils.py:
```diff
@@ -847,22 +818,18 @@ def load_module(name, value):
-                else:
-                    current_device_map = None
+            # 7.1 device_map shenanigans
+            current_device_map = _determine_current_device_map(device_map=final_device_map, component_name=name)
```
here too,  I think it is better without the method

sayakpaul replied:
Agreed. Done in d24465e04.

yiyixuxu commented on src/diffusers/pipelines/pipeline_utils.py:
```diff
@@ -900,68 +867,25 @@ def load_module(name, value):
-            passed_modules = set(list(init_kwargs.keys()) + list(passed_class_obj.keys())) - optional_kwargs
-            raise ValueError(
-                f"Pipeline {pipeline_class} expected {expected_modules}, but only {passed_modules} were passed."
+            init_kwargs = _update_init_kwargs_with_connected_pipeline(
```
very nice!

yiyixuxu reviewed (commented):
thanks! I left some feedback
Basically, the goal here is for someone not super familiar with the code to be able to glance through and follow the rough logic without having to go into these specific methods in a different file, so I think:
1. it is better not to add small functions with only a few lines of code that are already easy to read
2. add a little bit comments and provide an example of output of that function if it is not obvious

yiyixuxu commented on src/diffusers/pipelines/pipeline_loading_utils.py:
```diff
@@ -744,3 +745,157 @@ def _fetch_class_library_tuple(module):
+    connected_pipes = {prefix: getattr(modelcard.data, prefix, [None])[0] for prefix in CONNECTED_PIPES_KEYS}
+    load_kwargs = {
+        "cache_dir": pipeline_loading_kwargs.get("cache_dir"),
+        "force_download": pipeline_loading_kwargs.get("force_download"),
```
so for example `force_download` is default to `False` and same as `from_flax` etc https://github.com/huggingface/diffusers/blob/16a3dad474dad00f8e4071d699e1562471a2dacd/src/diffusers/pipelines/pipeline_utils.py#L626

and custom_pipeline is modified here https://github.com/huggingface/diffusers/blob/16a3dad474dad00f8e4071d699e1562471a2dacd/src/diffusers/pipelines/pipeline_utils.py#L742

so don't think pass the `kwargs.copy()` is the approach here

yiyixuxu replied:
in that case, we do not need to do anything, just pass `**pipeline_loading_kwargs`, no? cc @DN6 here too

sayakpaul replied:
Yeah exactly. Done in 7e811bfac.

sayakpaul replied:
But do also note this: 5519f8280

yiyixuxu commented on src/diffusers/pipelines/pipeline_loading_utils.py:
```diff
@@ -744,3 +745,157 @@ def _fetch_class_library_tuple(module):
+        return connected_passed_kwargs
+
+    connected_pipes = {
+        prefix: DiffusionPipeline.from_pretrained(repo_id, **load_kwargs.copy(), **get_connected_passed_kwargs(prefix))
```
```suggestion
        prefix: DiffusionPipeline.from_pretrained(repo_id, **pipeline_loading_kwargs, **get_connected_passed_kwargs(prefix))
```

I think this would work, no?

sayakpaul replied:
Ah right!

yiyixuxu commented on src/diffusers/pipelines/pipeline_utils.py:
```diff
@@ -900,68 +867,25 @@ def load_module(name, value):
+                **kwargs_copied,
             )
 
+        # 9. Potentially add passed objects if expected
```
I would also not add this method, because it contains important key logic one should not skip through here (also just a few lines)

sayakpaul commented on src/diffusers/pipelines/pipeline_loading_utils.py:
```diff
@@ -744,3 +745,157 @@ def _fetch_class_library_tuple(module):
+        "low_cpu_mem_usage": pipeline_loading_kwargs.get("low_cpu_mem_usage"),
+        "variant": pipeline_loading_kwargs.get("variant"),
+        "use_safetensors": pipeline_loading_kwargs.get("use_safetensors"),
+    }
```
Done in 7e811bfac.

sayakpaul commented on src/diffusers/pipelines/pipeline_utils.py:
```diff
@@ -724,59 +734,28 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
-                )
-                if variant_exists:
-                    model_variants[folder] = variant
+        model_variants = _identify_model_variants(folder=cached_folder, variant=variant, config=config_dict)
```
Done: e847f0f0f.

sayakpaul commented:
@yiyixuxu I have addressed your comments. 

I can do a clean up od `download()` following similar guidelines either in this PR or in a follow-up. LMK. Cc @DN6 too.

yiyixuxu reviewed (approved):
thanks!

sayakpaul commented:
@a-r-r-o-w just a heads up that I am going to merge this PR now. Perhaps you could follow the same principles while doing the refactoring of `from_pretrained()` of `ModelMixin`?

a-r-r-o-w commented on src/diffusers/pipelines/pipeline_loading_utils.py:
```diff
@@ -746,3 +747,92 @@ def _fetch_class_library_tuple(module):
+    model_variants = {}
+    if variant is not None:
+        for folder in os.listdir(folder):
+            folder_path = os.path.join(folder, folder)
```
Thanks for the notification, Sayak. I will work on refactoring `download()` now that this is merged.

I was trying to understand the changes in this PR, but this line feels like a regression from old behaviour:

```python
        # 2. Define which model components should load variants
        # We retrieve the information by matching whether variant
        # model checkpoints exist in the subfolders
        model_variants = {}
        if variant is not None:
            for folder in os.listdir(cached_folder):
                folder_path = os.path.join(cached_folder, folder)
                is_folder = os.path.isdir(folder_path) and folder in config_dict
                variant_exists = is_folder and any(
                    p.split(".")[1].startswith(variant) for p in os.listdir(folder_path)
                )
                if variant_exists:
                    model_variants[folder] = variant
```

In the old behaviour, we were doing `os.path.join(cached_folder, folder)`. But here, we are doing `os.path.join(folder, folder)`, where `folder` is the loop variable. It seems incorrect. WDYT? cc @DN6

yiyixuxu replied:
@a-r-r-o-w nice catch! can you fix?

a-r-r-o-w replied:
Thanks, I can open a PR shortly after understanding all the changes
